File: app/core/rate_limit.py
import time

import redis.asyncio as aioredis

# In-memory storage fallback if Redis is disabled or unavailable during tests
_in_memory_rate_limits: dict[str, list[float]] = {}


async def is_rate_limited(
    redis: aioredis.Redis | None,
    key: str,
    max_requests: int,
    window_seconds: int = 60,
) -> tuple[bool, int]:
    """
    Sliding window log rate limiter.
    Returns: (is_limited: bool, remaining_requests: int)
    """
    now = time.time()
    clear_before = now - window_seconds

    if redis is not None:
        try:
            pipe = redis.pipeline()
            pipe.zremrangebyscore(key, 0, clear_before)
            pipe.zadd(key, {str(now): now})
            pipe.zcard(key)
            pipe.expire(key, window_seconds + 5)
            results = await pipe.execute()
            current_count = results[2]

            remaining = max(0, max_requests - current_count)
            if current_count > max_requests:
                return True, 0
            return False, remaining
        except Exception:
            pass

    # In-memory fallback
    timestamps = _in_memory_rate_limits.get(key, [])
    timestamps = [t for t in timestamps if t > clear_before]
    timestamps.append(now)
    _in_memory_rate_limits[key] = timestamps

    current_count = len(timestamps)
    remaining = max(0, max_requests - current_count)
    if current_count > max_requests:
        return True, 0
    return False, remaining


# In-memory storage for temporary blocks when Redis is unavailable
_in_memory_blocked_phones: dict[str, float] = {}
_in_memory_blocked_customers: dict[str, float] = {}


async def block_phone_number(
    redis: aioredis.Redis | None,
    phone_hash: str,
    duration_seconds: int = 3600,
) -> None:
    """Temporarily blocks a phone number due to excessive failures or abuse."""
    key = f"phone:blocked:{phone_hash}"
    if redis is not None:
        try:
            await redis.set(key, "1", ex=duration_seconds)
            return
        except Exception:
            pass
    _in_memory_blocked_phones[phone_hash] = time.time() + duration_seconds


async def is_phone_blocked(
    redis: aioredis.Redis | None,
    phone_hash: str,
) -> bool:
    """Checks if a phone number is currently temporarily blocked."""
    key = f"phone:blocked:{phone_hash}"
    if redis is not None:
        try:
            exists = await redis.get(key)
            return bool(exists)
        except Exception:
            pass
    unblock_time = _in_memory_blocked_phones.get(phone_hash, 0.0)
    return time.time() < unblock_time


async def block_customer(
    redis: aioredis.Redis | None,
    customer_id: str,
    duration_seconds: int = 86400,
) -> None:
    """Temporarily blocks a customer account due to severe abuse detection."""
    key = f"customer:blocked:{customer_id}"
    if redis is not None:
        try:
            await redis.set(key, "1", ex=duration_seconds)
            return
        except Exception:
            pass
    _in_memory_blocked_customers[customer_id] = time.time() + duration_seconds


async def is_customer_blocked(
    redis: aioredis.Redis | None,
    customer_id: str,
) -> bool:
    """Checks if a customer account is currently blocked for abuse."""
    key = f"customer:blocked:{customer_id}"
    if redis is not None:
        try:
            exists = await redis.get(key)
            return bool(exists)
        except Exception:
            pass
    unblock_time = _in_memory_blocked_customers.get(customer_id, 0.0)
    return time.time() < unblock_time

File: src/middleware/rate-limit.ts
import { redis } from "../plugins/redis.js";

const inMemoryRateLimits = new Map<string, number[]>();
const inMemoryBlockedPhones = new Map<string, number>();
const inMemoryBlockedCustomers = new Map<string, number>();

/**
 * Sliding window log rate limiter using Redis ZSET with memory fallback.
 * Returns: { isLimited: boolean, remaining: number }
 */
export async function isRateLimited(
  key: string,
  maxRequests: number,
  windowSeconds = 60
): Promise<{ isLimited: boolean; remaining: number }> {
  const now = Date.now() / 1000;
  const clearBefore = now - windowSeconds;

  try {
    const pipeline = redis.pipeline();
    pipeline.zremrangebyscore(key, 0, clearBefore);
    pipeline.zadd(key, now, `${now}-${Math.random()}`);
    pipeline.zcard(key);
    pipeline.expire(key, windowSeconds + 5);
    const results = await pipeline.exec();

    if (results && results[2] && results[2][1] !== undefined) {
      const currentCount = Number(results[2][1]);
      const remaining = Math.max(0, maxRequests - currentCount);
      return {
        isLimited: currentCount > maxRequests,
        remaining,
      };
    }
  } catch {
    // Fall back to in-memory store
  }

  const timestamps = (inMemoryRateLimits.get(key) || []).filter((t) => t > clearBefore);
  timestamps.push(now);
  inMemoryRateLimits.set(key, timestamps);

  const currentCount = timestamps.length;
  const remaining = Math.max(0, maxRequests - currentCount);
  return {
    isLimited: currentCount > maxRequests,
    remaining,
  };
}

export async function blockPhoneNumber(phoneHash: string, durationSeconds = 3600): Promise<void> {
  const key = `phone:blocked:${phoneHash}`;
  try {
    await redis.set(key, "1", "EX", durationSeconds);
    return;
  } catch {
    inMemoryBlockedPhones.set(phoneHash, Date.now() / 1000 + durationSeconds);
  }
}

export async function isPhoneBlocked(phoneHash: string): Promise<boolean> {
  const key = `phone:blocked:${phoneHash}`;
  try {
    const exists = await redis.get(key);
    if (exists) return true;
  } catch {
    // Fall back
  }
  const unblockTime = inMemoryBlockedPhones.get(phoneHash) || 0;
  return Date.now() / 1000 < unblockTime;
}

export async function blockCustomer(customerId: string, durationSeconds = 86400): Promise<void> {
  const key = `customer:blocked:${customerId}`;
  try {
    await redis.set(key, "1", "EX", durationSeconds);
    return;
  } catch {
    inMemoryBlockedCustomers.set(customerId, Date.now() / 1000 + durationSeconds);
  }
}

export async function isCustomerBlocked(customerId: string): Promise<boolean> {
  const key = `customer:blocked:${customerId}`;
  try {
    const exists = await redis.get(key);
    if (exists) return true;
  } catch {
    // Fall back
  }
  const unblockTime = inMemoryBlockedCustomers.get(customerId) || 0;
  return Date.now() / 1000 < unblockTime;
}
